enhancements/etcd/scaling-etcd-with-raft-learners.md

```diff
@@ -0,0 +1,343 @@
+place in the cluster resulting in a new static pod revision. The `etcdenvvar controller` provides the etcd runtime
+environment variables to the `Target Config Controller` which are then is used to populate the etcd static pod manifest.
+
+`ETCD_INITIAL_CLUSTER` ENV variable is a critical part of etcd scaling process. Before a member has joined the cluster
+and received a snapshot from the leader including the cluster membership details from the member bucket the new etcd
+member must be able to communicate with its peers. This proposal includes moving population of
+`ETCD_INITIAL_CLUSTER` from etcd itself via `discover-etcd-initial-cluster`[1] to the `etcdenvvar controller` in the
+`cluster-etcd-operator`.
+
+[1] https://github.com/openshift/etcd/tree/openshift-4.10/openshift-tools/pkg/discover-etcd-initial-cluster
+
+### cluster-etcd-operator verify membership command
+
+In 4.9 the `verify` command was added to `cluster-etcd-operator` as way to ensure that before a backup is taken
+there is appropriate disk space available (verify storage). In order for cluster membership to be adjusted safely extra
+precautions and logical gates needs to be added. The logic for these extra safety steps will be added as a sub
+command `verify membership` and will be added as init container in the etcd static pod.
+
+The `verify membership` command will consume a new ENV `ETCD_MEMBER_IDS` populated by `etcdenvvar controller`. This
+ENV will take a similar map format as `ETCD_INITIAL_CLUSTER` <member_id>=<peerurls>. The verify command will ask the
+cluster for the current etcd members `MemberListRequest` and compare the member IDs expected in the ENV vs actual. The
+expectation is that any change in cluster membership is matched with a static pod revision containing the
+appropriate pairing.
+
+#### Reads
+
+1. `MemberListResponse`: The member list provides membership details on the etcd cluster.
+
+- Consumers: `etcdenvvar controller`, `verify membership`
+
+- ref: https://pkg.go.dev/github.com/coreos/etcd/etcdserver/etcdserverpb#MemberListResponse
+
+#### Writes
+
+1. `ETCD_INITIAL_CLUSTER`: This variable will be added to the etcd static pod ENV, allowing new etcd members to join
+the cluster.
+
+- Writer: new function added to `envVarFns`[1].
+
+- Consumer: etcd binary during first join of cluster, afterwards it is ignored.
+
+2. `ETCD_MEMBER_IDS`: This variable will be added to the etcd static pod EMV, the mapping of this data enables extra
+precautions during scaling.
+
+- Writer: new function added to `envVarFns`[1].
+
+- Consumer: cluster-etcd-operator verify membership command.
+
+New etcd members on an existing node will need to archive the data-dir of the previous member to allow scale up to
+occur. `verify membership` can observe this by asking the cluster for membership details and using the conditions of
+`Name=""` and the method of `Member` `IsLearner` to conclude this.
```

Comment:
Is this supported? I was under the impression that we had a 1:1 mapping of etcd member to instance, and that, if removed from the cluster, we expect the node to go away. This could be a misunderstanding on my part, but if this is a change in behaviour from today I think this needs some more explanation on motivation and the risks around this change